Source/WebKit/UIProcess/Automation/WebAutomationSession.cpp

```diff
@@ -990,13 +988,138 @@ void WebAutomationSession::fragmentNavigatedForFrame(const WebFrameProxy& frame,
 {
     m_bidiProcessor->browsingContextDomainNotifier().fragmentNavigated(handleForWebFrameProxy(frame), navigationIDToProtocolString(navigationID), WallTime::now().secondsSinceEpoch().milliseconds(), frame.url().string());
 }
+
+void WebAutomationSession::contextCreatedForPage(const WebPageProxy& page)
+{
+    if (RefPtr mutablePage = WebProcessProxy::webPage(page.identifier())) {
+        mutablePage->getAllFrameTrees([this, protectedThis = Ref { *this }, pageID = page.identifier()](Vector<FrameTreeNodeData>&& trees) {
+            RefPtr page = WebProcessProxy::webPage(pageID);
+            if (!page)
+                return;
+
+            for (auto& tree : trees)
+                recursivelyEmitContextCreated(tree, std::nullopt);
+        });
+    }
+}
+
+std::pair<String, String> WebAutomationSession::getClientWindowAndUserContext(const WebPageProxy& page)
```

Comment:
This can be a static helper function rather than an instance method, if it's not being used outside the class.